### PR TITLE
feat(payments): Add inabox test coverage for payments system

### PR DESCRIPTION
**Pull request by litt3**
- Creates inabox payments tests
- Makes necessary changes to inabox to support the new tests

NOTE:
- Currently, none of the new payments tests run automatically. They take a long time, and cannot be run in parallel yet
- Eventually, once inabox pain points are resolved, these tests should be run in CI. Until then, at least we can get manual coverage.

### CODE REVIEW

File: inabox/tests/payments/reservation_only_test.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+	"github.com/Layr-Labs/eigenda/core/payments/clientledger"
+	"github.com/Layr-Labs/eigenda/core/payments/reservation"
+	integration "github.com/Layr-Labs/eigenda/inabox/tests"
+	"github.com/Layr-Labs/eigenda/test"
+	"github.com/Layr-Labs/eigenda/test/random"
+	"github.com/stretchr/testify/require"
+)
+
+// NOTE: Currently, it doesn't work to run these tests in sequence. Each test must be run as a separate command.
+// The problem is that the cleanup logic sometimes randomly fails to free docker ports, so subsequent setups fail.
```
Comment (dmanc) [inline inabox/tests/payments/reservation_only_test.go:18]: Is it possible to run the tests sequentially but with the same inabox deployment (no cleanup)?

Comment (litt3) [reply inabox/tests/payments/reservation_only_test.go:18]: The problem is that we need to configure the system in different ways: the point of these tests is to check different combinations of controller + client configurations

If we could redeploy individual parts of the deployment, e.g. keep all components except redeploy a new controller, then this could be much faster, and could be done without any cleanup between tests.

Since deployment is all-or-nothing for now, this was the only way I could think to accomplish the needed behavior

Comment (dmanc) [reply inabox/tests/payments/reservation_only_test.go:18]: I see thanks for the context. Redeployment sounds like a good feature but I think in the short term we would just run them in parallel once we can spin up inabox with OS allocated ports.

File: inabox/tests/payments/reservation_only_test.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+// Once we figure out why resources aren't being freed, then these tests will be runnable the "normal" way.
+
+func TestReservationOnly_LegacyClient_LegacyController(t *testing.T) {
```
Comment (dmanc) [inline inabox/tests/payments/reservation_only_test.go:21]: Might want to add something like 	t.Skip("Manual test for now"), not sure if the CI is going to try to run the tests automatically

File: contracts/script/EigenDADeployer.s.sol [hunk 3]
```diff
@@ -93,7 +93,7 @@ contract EigenDADeployer is DeployOpenEigenLayer {
     uint64 _pricePerSymbol = 0.447 gwei;
     uint64 _priceUpdateCooldown = 1;
     uint64 _globalSymbolsPerPeriod = 131072;
-    uint64 _reservationPeriodInterval = 300;
+    uint64 _reservationPeriodInterval = 10;
```
Comment (samlaf) [inline contracts/script/EigenDADeployer.s.sol:96]: Should we move this script to some contracts/script/inabox/ folder?

File: contracts/script/SetUpEigenDA.s.sol [hunk 4]
```diff
@@ -158,7 +158,7 @@ contract SetupEigenDA is EigenDADeployer, EigenLayerUtils {
 
         // Register Reservations for client as the eigenDACommunityMultisig
         IPaymentVault.Reservation memory reservation = IPaymentVault.Reservation({
-            symbolsPerSecond: 452198,
+            symbolsPerSecond: uint64(vm.envOr("USER_RESERVATION_SYMBOLS_PER_SECOND", uint256(452198))),
```
Comment (samlaf) [inline contracts/script/SetUpEigenDA.s.sol:161]: You know how much I hate env vars. There is a config file for this script afair, we could add it there instead?

Or even better do we have golang code to setup a reservation? I feel like we should setup reservations from golang code, then we can set them up whenever/however we wish in your golang tests.

File: contracts/script/SetUpEigenDA.s.sol [hunk 5]
```diff
@@ -167,8 +167,6 @@ contract SetupEigenDA is EigenDADeployer, EigenLayerUtils {
         address clientAddress = address(0x1aa8226f6d354380dDE75eE6B634875c4203e522);
         vm.startBroadcast(msg.sender);
         paymentVault.setReservation(clientAddress, reservation);
-        // Deposit OnDemand
-        paymentVault.depositOnDemand{value: 0.1 ether}(clientAddress);
```
Comment (samlaf) [inline contracts/script/SetUpEigenDA.s.sol:171]: why remove this?

File: core/thegraph/state_integration_test.go [hunk 6]
```diff
@@ -47,6 +47,7 @@ func setupTest(t *testing.T) *inaboxtests.InfrastructureHarness {
 		TestName:          testName,
 		InMemoryBlobStore: true, // Graph indexer test doesn't need blob store
 		Logger:            logger,
+		RootPath:          "../../",
```
Comment (samlaf) [inline core/thegraph/state_integration_test.go:50]: ehh we really shouldn't be doing this! This is so error prone. If you ever run this test from some other folder for whatever reason (like from root) it'll break.

@dmanc we really need to fix these paths issues. The right way is to have a file either in root or in some common/paths package, read where it is using golang stdlib, then path.join(../..) to it, and make that a global var that other packages can import. Then regardless of where the package live, or where the code is executed, that RootPath variable will ALWAYS be valid.

File: disperser/apiserver/server_v2.go [hunk 7]
```diff
@@ -143,7 +143,7 @@ func NewDispersalServerV2(
 		controllerConnection = connection
 		controllerClient = controller.NewControllerServiceClient(connection)
 
-		logger.Info("Using controller-based payment system")
+		logger.Info("Using controller-based payment system", "controller", controllerAddress)
```
Comment (samlaf) [inline disperser/apiserver/server_v2.go:146]: nit which def doesnt belong in this PR: we should move away from using "address" to mean URL. Whenever I see address I think of ethereum address. We should use "host" and "port" and "url" for rpc "addresses".

File: inabox/Makefile [hunk 8]
```diff
@@ -1,10 +1,30 @@
...
+run-payments-tests:
+	@for test in \
+		TestReservationOnly_LegacyClient_LegacyController \
+		TestReservationOnly_LegacyClient_NewController \
+		TestReservationOnly_NewClient_LegacyController \
+		TestReservationOnly_NewClient_NewController; do \
+		echo "Running $$test..."; \
+		if ! go test ./tests/payments -v -run "^$$test$$"; then \
+			echo "❌ TEST FAILED: $$test"; \
+			$(MAKE) -s stop-services stop-infra; \
```
Comment (samlaf) [inline inabox/Makefile]: nit: you can use `stop-inabox` to stop both infra and services

File: inabox/deploy/config.go [hunk 9]
```diff
@@ -221,12 +221,20 @@ func (env *Config) generateDisperserV2Vars(ind int, logPath, dbPath, grpcPort st
 		DISPERSER_SERVER_RESERVATIONS_TABLE_NAME: "e2e_v2_reservation",
 		DISPERSER_SERVER_ON_DEMAND_TABLE_NAME:    "e2e_v2_ondemand",
 		DISPERSER_SERVER_GLOBAL_RATE_TABLE_NAME:  "e2e_v2_global_reservation",
+		DISPERSER_SERVER_CONTROLLER_ADDRESS:      "localhost:30000",
```
Comment (samlaf) [inline inabox/deploy/config.go]: we need to move away from hardcoded ports, but at least for now could you make it a const at the top of the file? I see it reused in at least 2 places.

File: inabox/tests/setup_infra.go [hunk 10]
```diff
@@ -5,20 +5,25 @@ import (
...
+	TestName                        string
+	InMemoryBlobStore               bool
+	Logger                          logging.Logger
+	RootPath                        string
+	MetadataTableName               string
+	BucketTableName                 string
+	MetadataTableNameV2             string
+	UserReservationSymbolsPerSecond uint64
+	ClientLedgerMode                clientledger.ClientLedgerMode
+	ControllerUseNewPayments        bool
```
Comment (samlaf) [inline inabox/tests/setup_infra.go:33]: yikes I can see this getting out of hands again.
This was meant to be an infra only config. But we're now questioning what "infra" means, given that thegraph and localstack are actually disperser-specific infra (potentially, although not sure its needed for inabox...)

In any case these disperser specific config stuff should live in some other ServicesConfig or something. Haven't figured out the right organization but want to avoid these bloated 40+fields configs.

File: inabox/tests/payments/payload_submitter.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+	"github.com/Layr-Labs/eigenda/api/clients/v2/coretypes"
+	"github.com/Layr-Labs/eigenda/api/clients/v2/payloaddispersal"
+	"github.com/Layr-Labs/eigenda/test/random"
+)
+
+// Submits payloads at a certain rate for a duration.
+//
+// Returns dispersal errors on a channel.
+func SubmitPayloads(
```
Comment (samlaf) [inline inabox/tests/payments/payload_submitter.go]:
```suggestion
func submitPayloads(
```

Comment (samlaf) [reply inabox/tests/payments/payload_submitter.go]: Q: I know that this is very simple, but could it reuse the load generator code instead?

File: inabox/tests/setup_test_harness.go [hunk 12]
```diff
@@ -117,10 +127,14 @@ func NewTestHarnessWithSetup(infra *InfrastructureHarness) (*TestHarness, error)
 	}
 
 	// Setup payload disperser
-	if err := setupPayloadDisperserForContext(testCtx, infra); err != nil {
+	if err := setupPayloadDisperserForContext(ctx, testCtx, infra); err != nil {
```
Comment (samlaf) [inline inabox/tests/setup_test_harness.go:130]: ok so imo (havent checked all the dependencies and whether this works cleanly given current state of inabox), clients should NOT be part of the inabox infra. They should be built on a per-test basis when needed.

This is the right separation of concern imo. This way you can control your payment tests, especially parallelization. Run 4 tests in parallel, they each start by creating a new reservation, and then build the clients using that private key.

File: inabox/tests/setup_test_harness.go [hunk 13]
```diff
@@ -362,3 +553,30 @@ func newTransactOptsFromPrivateKey(privateKeyHex string, chainID *big.Int) *bind
 
 	return opts
 }
+
+func setupPaymentVaultTransactor(
```
Comment (samlaf) [inline inabox/tests/setup_test_harness.go:557]: this is also a very generic/useful component right, that we could reuse as part of a CLI tool? Why have this under inabox?

File: inabox/tests/payments/reservation_only_test.go [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+
+	infra, err := integration.SetupInfrastructure(infraConfig)
+	require.NoError(t, err)
+
+	testHarness, err := integration.NewTestHarnessWithSetup(infra)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		testHarness.Cleanup()
+		integration.TeardownInfrastructure(infra)
```
Comment (samlaf) [inline inabox/tests/payments/reservation_only_test.go]: you prob want to call teardownInfra right after SetupInfra, otherwise if NewTestHarnessWithSetup errors, infra will not get cleaned.

File: inabox/tests/payments/reservation_only_test.go [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+
+	t.Run("Within reservation limits", func(t *testing.T) {
+		initialReservation, err := reservation.NewReservation(
+			1024,
+			time.Now().Add(-1*time.Hour),
+			time.Now().Add(24*time.Hour),
+			[]core.QuorumID{0, 1},
+		)
+		require.NoError(t, err)
+		err = testHarness.UpdateReservation(t.Context(), t, initialReservation)
```
Comment (samlaf) [inline inabox/tests/payments/reservation_only_test.go]: nit: can we rename `UpdateReservationOnchain` or something to make it obvious we're making an eth tx? otherwise it could read as "read from onchain and update the test harness' reservation"

File: inabox/tests/payments/reservation_only_test.go (hunk 2 above)
Comment (litt3) [reply inabox/tests/payments/reservation_only_test.go:21]: d21cb0cbd1b8a2fc725f70de55467d7e56265558

File: contracts/script/SetUpEigenDA.s.sol (hunk 4 above)
Comment (litt3) [reply contracts/script/SetUpEigenDA.s.sol:161]: > You know how much I hate env vars

I share your dislike, but I'm not sure there's a better solution in the short term.

> There is a config file for this script afair, we could add it there instead?

The problem is that I need to define this at test time. So short of duplicating the config file with a single value changed, I'm not sure how this would be accomplished. Even if we had duplicate config files, there would have to be some mechanism of injecting the correct config file path to this deployment script, and I'm not sure how else we'd do it, other than env vars. 

> Or even better do we have golang code to setup a reservation? I feel like we should setup reservations from golang code, then we can set them up whenever/however we wish in your golang tests.

I wanted to go that way, but it's not trivial to do based on how the infrastructure is all deployed together. The flow we would _want_ is this:

1. deploy infra
2. submit some smart contract transactions
3. spin up eigenDA components

This is further complicated by the fact that the old `Accountant` doesn't observe `PaymentVault` updates in real time, so the reservation must be in place before the clients are created.

File: contracts/script/SetUpEigenDA.s.sol (hunk 5 above)
Comment (litt3) [reply contracts/script/SetUpEigenDA.s.sol:171]: I need to test edge cases which include NOT having enough on demand funds. This is a huge deposit, and it's not possible to un-deposit funds at test time.

I removed this, and we can just deposit on demand funds as needed, on a test by test bases. Most tests will probably rely on reservations, anyway.

File: core/thegraph/state_integration_test.go (hunk 6 above)
Comment (litt3) [reply core/thegraph/state_integration_test.go:50]: > ehh we really shouldn't be doing this! This is so error prone.

💯 

Current code had this hardcoded, which made it impossible to put tests in a subdir. This is just my workaround to get these tests written, without overhauling inabox right away

File: disperser/apiserver/server_v2.go (hunk 7 above)
Comment (litt3) [reply disperser/apiserver/server_v2.go:146]: I feel like it's generally clear from context, but I agree it's not ideal.

> We should use "host" and "port" and "url" for rpc "addresses".

Makes sense, I'm on board with that. As long as we don't try to use `socket` 🤮

File: inabox/tests/payments/payload_submitter.go (hunk 11 above)
Comment (litt3) [reply inabox/tests/payments/payload_submitter.go]: > could it reuse the load generator code instead?

I took a look- I don't think that could be reused here with very substantial modification. The load generator is doing a ton of stuff, and here we only need to do a very small fraction.

It's also nice to be able to add logs without affecting other things, and to have access to a `testing.T` instance, which we wouldn't want to add to the load generator

File: inabox/Makefile (hunk 8 above)
Comment (litt3) [reply inabox/Makefile]: 7c86143fe9d43c8c0d6082fe6f23f2a0dcb7db60

File: inabox/tests/setup_infra.go (hunk 10 above)
Comment (litt3) [reply inabox/tests/setup_infra.go:33]: I anticipated some discussion on this point. Some things I considered:

- All three of these fields are temporary: once the old payment system is removed, they will be deleted
- Each of these three values belongs to a different component. So if I put these into specific config structs, and nested those inside of the infra config, we would still have 3 fields here. But each field would be a config struct that contains only a single value. 
- I think the correct way to do this would be to have separate setup steps for each part of the infrastructure stack. So there would be a `ControllerConfig`, but not nested inside the `InfraConfig`. It would be totally separate, passed into some kind of `SetupController` function. Since reworking the flow of instantiation is a much larger change, I opted to keep it simple for now.

File: contracts/script/EigenDADeployer.s.sol (hunk 3 above)
Comment (litt3) [reply contracts/script/EigenDADeployer.s.sol:96]: Makes sense to me (but in a separate PR)

File: inabox/deploy/config.go (hunk 9 above)
Comment (litt3) [reply inabox/deploy/config.go]: > we need to move away from hardcoded ports

👍 

Not sure what exactly that would look like in this PR. So for now:
9a0ba5f1648e15d9cb27d1108c8330e9d0f97be9

File: inabox/tests/payments/payload_submitter.go (hunk 11 above)
Comment (litt3) [reply inabox/tests/payments/payload_submitter.go]: 964ff49b7eded5c9461ee5b53ee674d728c3b735

edit:
whoops, missed some changes. 87dcf91655d89f014d5d4507b669ed85df0c241a

File: inabox/tests/setup_test_harness.go (hunk 12 above)
Comment (litt3) [reply inabox/tests/setup_test_harness.go:130]: Definitely on board. 

As I've mentioned elsewhere, I feel like our infra setup should be split in to multiple parts _in general_. But it's especially the case for clients, that they shouldn't be part of the "infra".

I don't think this will be hugely difficult to pull off, but there will be some level of reorganization, which I think exceeds the amount of inabox tweaks that fit into this PR.

File: inabox/tests/setup_test_harness.go (hunk 13 above)
Comment (litt3) [reply inabox/tests/setup_test_harness.go:557]: The transactor itself is provided by the bindings. This method is just to instantiate the existing transactor, so that tests can access it through the `TestHarness`. 

We might want a CLI to update the `PaymentVault` at some point, but this method here isn't directly applicable to anything beyond the scope of the `TestHarness`

File: inabox/tests/payments/reservation_only_test.go (hunk 14 above)
Comment (litt3) [reply inabox/tests/payments/reservation_only_test.go]: 4735210e6558883977655f0695e953d0f9a9b1a7

File: inabox/tests/payments/reservation_only_test.go (hunk 15 above)
Comment (litt3) [reply inabox/tests/payments/reservation_only_test.go]: 30e6d8a2b5085ce5b515ff30845d3cb46c6730e6

**Review by dmanc — APPROVED** [review]

File: contracts/script/SetUpEigenDA.s.sol (hunk 5 above)
Comment (pakim249CAL) [reply contracts/script/SetUpEigenDA.s.sol:171]: I think a deployment generally should NOT be setting stuff like this, when it can be accomplished with a regular tx. So I'm not against this change.

**Review by pakim249CAL — DISMISSED** [review]

File: contracts/script/SetUpEigenDA.s.sol (hunk 4 above)
Comment (samlaf) [reply contracts/script/SetUpEigenDA.s.sol:161]: @dmanc here's a good reason for separating infra deployment (at least chain) from eigenda services!

File: contracts/script/SetUpEigenDA.s.sol (hunk 5 above)
Comment (samlaf) [reply contracts/script/SetUpEigenDA.s.sol:171]: Issue is that this is used in inabox the devnet environment, where we most likely do want a deposit.

@litt3 do you have an example test where you do a deposit offchain? Didn't see one.

@pakim249CAL this is a setup script though, not a deployment script. It first deploys Eiegenlayer and EigenDA by calling some other script. And then it registers operators and clients. Tbh I agree we should have offchain code to setup operators/clients, but until we do, I prefer we keep this script working for devnet.

File: core/thegraph/state_integration_test.go (hunk 6 above)
Comment (samlaf) [reply core/thegraph/state_integration_test.go:50]: Oh you introduced RootPath in this PR? If it was already hardcoded, then it was probably hardcoded as a relative path, which was the issue. Could you just change that hardcoded path to be an absolute path using the strategy I suggested above. I'd prefer we don't introduce more custom relative path logic like this.

File: inabox/tests/setup_infra.go (hunk 10 above)
Comment (samlaf) [reply inabox/tests/setup_infra.go:33]: agree with your 3rd bullet point, that's also how I'm envisioning it. In the meantime if this is just temporary stuff could you add a "legacy payment stuff; will soon be deprecated" comment above those 3 fields and keep them separate from the rest?

File: contracts/script/SetUpEigenDA.s.sol (hunk 5 above)
Comment (litt3) [reply contracts/script/SetUpEigenDA.s.sol:171]: > Issue is that this is used in inabox the devnet environment, where we most likely do want a deposit.

Why would anyone care to have an on-demand deposit? There is a massive ~16MiB/sec default reservation, which can cover any conceivable rate of dispersal. As currently configured, this default on-demand deposit has never once actually been used. I don't see any reason to use on-demand payments for tests, other than for cases where you're specifically testing payment behavior.

> do you have an example test where you do a deposit offchain? Didn't see one.

Additional tests are coming for on-demand. The tests in this PR are only for reservation based payments.

That aside, even these reservation-only tests rely on _not_ having a huge on-demand deposit for the test account. These tests make assertions about failed dispersals when the reservation cannot cover a dispersal, so having a big on-demand deposit removes the ability to make any assertions about the limits of a reservation (for old clients, that don't declare payment mode).

File: inabox/tests/setup_infra.go (hunk 10 above)
Comment (litt3) [reply inabox/tests/setup_infra.go:33]: 27f08b8a869385d0ca0d8319c5bb3e2858830829

File: core/thegraph/state_integration_test.go (hunk 6 above)
Comment (litt3) [reply core/thegraph/state_integration_test.go:50]: > Could you just change that hardcoded path to be an absolute path using the strategy I suggested above

I implemented this, and it's too big to include in this PR. There are _lots_ of touch points with root path within inabox.

> I'd prefer we don't introduce more custom relative path logic like this.

This isn't introducing any custom logic, it's just the existing logic but slightly expanded to support different nesting of the test location.

I'm on board with the change in general, but it's not something that belongs tacked onto this PR

File: contracts/script/SetUpEigenDA.s.sol (hunk 5 above)
Comment (samlaf) [reply contracts/script/SetUpEigenDA.s.sol:171]: I see. makes sense. resolving

**Review by samlaf — APPROVED** [review]
